matchmaking/matchmaker.py

```diff
@@ -0,0 +1,344 @@
+                boost += 1
+        return boost
+
+    def _penalty_based_on_recent_pairing(self, profile1, profile2):
+        # if they've met too recently, then score is zero and fast return
+        # if they've never met, there is no penalty
+        # if they have met, then penalize score based on recency (max penalty = 1.0)
+        penalty = 0
+        p1p2_key = key(profile1, profile2)
+        recent_match_by_profile_pair = None
+        if p1p2_key in self._recent_match_by_profile_pair:
+            recent_match_by_profile_pair = self._recent_match_by_profile_pair[p1p2_key]
+            time_since_last_meeting = self._now - recent_match_by_profile_pair['date']
+            if time_since_last_meeting < self._cutoff_for_repairing:
+                penalty = float('infinity')
+            else:
+                # if time_since_last_meeting == self._cutoff_for_repairing
+                #   then the penalty is 1.0
+                # if time_since_last_meeting == self._cutoff_for_repairing + self._repairing_penalty_halflife
+                #   then the penalty is 0.5
+                # if time_since_last_meeting == self._cutoff_for_repairing + 2*self._repairing_penalty_halflife
+                #   then the penalty is 0.25
+                penalty = 2 ** (-time_since_last_meeting / self._repairing_penalty_halflife)
+        return penalty
+
+    def _select_topic(self, profile1, profile2, met_recently, topics):
+        # if there are multiple topics they could be matched in, then weed out any where they've already met (less
+        # recently than above)
+        topics_this_pair_has_discussed = set()
+        least_recent_topic = None
+        least_recent_date = None
+        for topic in topics:
+            p1p2top_key = key(profile1, profile2, topic)
+            if met_recently and p1p2top_key in self._recent_match_by_profile_pair_and_topic:
+                topics_this_pair_has_discussed.add(topic)
+                date_of_that_meeting = self._recent_match_by_profile_pair_and_topic[p1p2top_key]['date']
+                if least_recent_date is None or least_recent_date > date_of_that_meeting:
+                    least_recent_topic = topic
+                    least_recent_date = date_of_that_meeting
+        if len(topics - topics_this_pair_has_discussed) == 0:
+            # they've discussed all these topics, so just return the least recently discussed topic
+            topic = least_recent_topic
+        else:
+            # exclude topics they're already talked about and then just pick one
+            topics -= topics_this_pair_has_discussed
+
+            # TODO: here we're just picking one of the remaining topics but we could do better by selecting topics
+            # that at least one of them haven't been matched in before. And if that didn't uniquely identify a best topic
+            # we could find the topic where they've met least recently.
+            topic = topics.pop()
```

Comment:
I think for now you should at least `shuffle` the topics, just to add a bit of randomness. Otherwise, I _think_ that a person will talk about the same topic every time with a match they haven't had with before, because the topic channels should always be in the same order.

Reply:
wooooooo - good eye :D  fixed in 466af7f